src/core/address_space.cpp

```diff
@@ -244,22 +274,72 @@ enum PosixPageProtection {
+            mmap(reinterpret_cast<void*>(SYSTEM_RESERVED_MIN), system_reserved_size,
+                 PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS | MAP_NORESERVE, -1, 0));
+        user_base = reinterpret_cast<u8*>(mmap(reinterpret_cast<void*>(USER_MIN), user_size,
+                                               PROT_READ | PROT_WRITE,
+                                               MAP_PRIVATE | MAP_ANONYMOUS | MAP_NORESERVE, -1, 0));
+#else
+        const auto virtual_size = system_managed_size + system_reserved_size + user_size;
+        const auto virtual_base = reinterpret_cast<u8*>(
+            mmap(reinterpret_cast<void*>(SYSTEM_MANAGED_MIN), virtual_size, PROT_READ | PROT_WRITE,
                  MAP_PRIVATE | MAP_ANONYMOUS | MAP_NORESERVE | MAP_FIXED, -1, 0));
-        if (virtual_base == MAP_FAILED) {
+        system_managed_base = virtual_base;
+        system_managed_base = virtual_base + (SYSTEM_RESERVED_MIN - SYSTEM_MANAGED_MIN);
+        user_base = virtual_base + (USER_MIN - SYSTEM_MANAGED_MIN);
+#endif
+        if (system_managed_base == MAP_FAILED || system_reserved_base == MAP_FAILED ||
+            user_base == MAP_FAILED) {
             LOG_CRITICAL(Kernel_Vmm, "mmap failed: {}", strerror(errno));
             throw std::bad_alloc{};
         }
+
+        LOG_INFO(Kernel_Vmm, "System managed virtual memory region: {} - {}",
+                 fmt::ptr(system_managed_base),
+                 fmt::ptr(system_managed_base + system_managed_size - 1));
+        LOG_INFO(Kernel_Vmm, "System reserved virtual memory region: {} - {}",
+                 fmt::ptr(system_reserved_base),
+                 fmt::ptr(system_reserved_base + system_reserved_size - 1));
+        LOG_INFO(Kernel_Vmm, "User virtual memory region: {} - {}", fmt::ptr(user_base),
+                 fmt::ptr(user_base + user_size - 1));
+
+        const VAddr system_managed_addr = reinterpret_cast<VAddr>(system_managed_base);
+        const VAddr system_reserved_addr = reinterpret_cast<VAddr>(system_managed_base);
+        const VAddr user_addr = reinterpret_cast<VAddr>(user_base);
+        m_free_regions.insert({system_managed_addr, system_managed_addr + system_managed_size});
+        m_free_regions.insert({system_reserved_addr, system_reserved_addr + system_reserved_size});
+        m_free_regions.insert({user_addr, user_addr + user_size});
+
+#ifdef __APPLE__
+        const auto shm_path = fmt::format("/BackingDmem{}", getpid());
+        backing_fd = shm_open(shm_path.c_str(), O_RDWR | O_CREAT | O_EXCL, 0600);
+        if (backing_fd < 0) {
+            LOG_CRITICAL(Kernel_Vmm, "shm_open failed: {}", strerror(errno));
+            throw std::bad_alloc{};
+        }
+        shm_unlink(shm_path.c_str());
+#else
         madvise(virtual_base, virtual_size, MADV_HUGEPAGE);
 
+        const VAddr start_addr = reinterpret_cast<VAddr>(virtual_base);
+        m_free_regions.insert({start_addr, start_addr + virtual_size});
```

Comment:
Isn't this the same as the insertion of system_managed_addr above?

Reply:
Accidentally got left in, thanks